### PR TITLE
ENH auto-detect backend based on PATH

### PR DESCRIPTION
**Pull request by ogrisel**
Try to look-up SLURM commands, then SGE commands in the PATH by default. Fail with an explicit error message otherwise.

### CODE REVIEW

**Comment by ogrisel** [comment]
@arjoly travis is green for this one. I will rebase my `futures` branch on master as soon as this is green.

**Comment by arjoly** [comment]
@ogrisel Could you wait a bit? I have found some last bugs that are not caught, because now everything run with slurm installed by default.

**Comment by arjoly** [comment]
A pull request is coming.

**Comment by ogrisel** [comment]
Ok let me know if you need me to rebase this branch on a future version of master.

File: clusterlib/scheduler.py [hunk 1]
```diff
@@ -24,6 +24,52 @@
...
+    else:
+        for path in os.environ["PATH"].split(os.pathsep):
+            path = path.strip('"')
+            exe_file = os.path.join(path, program)
+            if is_exe(exe_file):
+                return exe_file
+    return None
+
+
+def _get_backend(backend="auto"):
```
Comment (arjoly) [inline clusterlib/scheduler.py]: nice, i wanted to implement this!

File: clusterlib/scheduler.py [hunk 2]
```diff
@@ -24,6 +24,52 @@
...
+    """Detect the backend to use based on the commands present in the PATH"""
+    backend_commands = (
+        ('slurm', 'squeue'),
+        ('sge', 'qstat'),
+    )
+    if backend == "auto":
+        # If backend is auto, check that it is not configured explicitly
+        # in a dedicated environment variable
+        backend = os.environ.get('CLUSTERLIB_BACKEND', 'auto').lower()
```
Comment (arjoly) [inline clusterlib/scheduler.py]: Could you document this?

File: clusterlib/scheduler.py [hunk 3]
```diff
@@ -24,6 +24,52 @@
 ]
 
 
+def _which(program):
```
Comment (arjoly) [inline clusterlib/scheduler.py]: Why not using [shutil.which](https://docs.python.org/3.3/library/shutil.html#shutil.which) in python 3?

**Comment by arjoly** [comment]
Travis is now configured to run tests with and without slurm. 

This pull request could be merged if tests are provided.

File: clusterlib/scheduler.py (hunk 3 above)
Comment (ogrisel) [inline clusterlib/scheduler.py]: Indeed.

**Comment by ogrisel** [comment]
All your comments should have been addressed. Let's now wait for travis.

**Comment by ogrisel** [comment]
@arjoly travis is happy with and without SLURM. Any further comments?

File: clusterlib/scheduler.py [hunk 4]
```diff
@@ -146,8 +198,10 @@ def submit(job_command, job_name="job", time="24:00:00", memory=4000,
     log_directory : str, optional (default=None)
         Specify the log directory. If None, no log directory is specified.
 
-    backend : {'sge', 'slurm'}, optional (default="slurm")
-        Backend where the job will be submitted
+    backend : {'auto', 'slurm', 'sge'}, optional (default="auto")
+        Backend where the job will be submitted. If 'auto', try detect
+        the backend to use based on the commands available in the PATH
+        variable.
```
Comment (arjoly) [inline clusterlib/scheduler.py]: Can you mention `CLUSTERLIB_BACKEND ` here?

Comment (ogrisel) [inline clusterlib/scheduler.py]: Indeed.

**Comment by arjoly** [comment]
Can you add your contribution in doc/whats_new.rst?

**Comment by arjoly** [comment]
When remaining remarks are addressed, I will merge this. Great job !!! :-) Thanks a lot @ogrisel

**Comment by ogrisel** [comment]
Done.

**Comment by arjoly** [comment]
Merged by rebase !

**Comment by arjoly** [comment]
:beers:

**Comment by ogrisel** [comment]
Thanks! :beers:

**Pull request merged**